core/src/main/scala/io/finch/endpoint/param.scala

```diff
@@ -1,80 +1,94 @@
+  final def apply(input: Input): Endpoint.Result[F[A]] =
     input.request.params.get(name) match {
       case None => missing(input, name)
-      case Some(value) => present(input, value)
+      case Some(value) =>
+        d(value) match {
+          case Return(r) => present(input, r)
+          case Throw(e) => EndpointResult.Matched(input, Rs.paramNotParsed(name, tag, e))
+        }
     }
 
   final override def item: items.RequestItem = items.ParamItem(name)
   final override def toString: String = s"param($name)"
+
+
 }
 
 private object Param {
-  trait Required { _: Param[String] =>
-    protected def missing(input: Input, name: String): Endpoint.Result[String] =
+  trait Required[A] { _: Param[Id, A] =>
+    protected def missing(input: Input, name: String): Endpoint.Result[A] =
       EndpointResult.Matched(input, Rs.paramNotPresent(name))
 
-    protected def present(input: Input, value: String): Endpoint.Result[String] =
+    protected def present(input: Input, value: A): Endpoint.Result[A] =
       EndpointResult.Matched(input, Rs.payload(value))
   }
 
-  trait Optional { _: Param[Option[String]] =>
-    protected def missing(input: Input, name: String): Endpoint.Result[Option[String]] =
+  trait Optional[A] { _: Param[Option, A] =>
+    protected def missing(input: Input, name: String): Endpoint.Result[Option[A]] =
       EndpointResult.Matched(input, Rs.none)
 
-    protected def present(input: Input, value: String): Endpoint.Result[Option[String]] =
+    protected def present(input: Input, value: A): Endpoint.Result[Option[A]] =
       EndpointResult.Matched(input, Rs.payload(Some(value)))
   }
 
-  trait Exists { _: Param[String] =>
-    protected def missing(input: Input, name: String): Endpoint.Result[String] =
+  trait Exists[A] { _: Param[Id, A] =>
+    protected def missing(input: Input, name: String): Endpoint.Result[A] =
       EndpointResult.Skipped
 
-    protected def present(input: Input, value: String): Endpoint.Result[String] =
+    protected def present(input: Input, value: A): Endpoint.Result[A] =
       EndpointResult.Matched(input, Rs.payload(value))
   }
 }
 
-private abstract class Params[A](name: String) extends Endpoint[A] {
+private abstract class Params[F[_], A](name: String, d: DecodeEntity[A], tag: ClassTag[A]) extends Endpoint[F[A]] {
 
-  protected def missing(input: Input, name: String): Endpoint.Result[A]
-  protected def present(input: Input, value: Iterable[String]): Endpoint.Result[A]
+  protected def missing(input: Input, name: String): Endpoint.Result[F[A]]
+  protected def present(input: Input, value: Iterable[A]): Endpoint.Result[F[A]]
 
-  def apply(input: Input): Endpoint.Result[A] = input.request.params.getAll(name) match {
+  def apply(input: Input): Endpoint.Result[F[A]] = input.request.params.getAll(name) match {
     case value if value.isEmpty => missing(input, name)
-    case value => present(input, value)
+    case value => Try.collect(value.map(d.apply).toSeq) match {
```

Comment:
I think we should keep collecting the errors here. Right now we return just the first observed failure. See `StringNelEndpointOps` for an inspiration.